ibis/expr/operations.py

```diff
@@ -483,12 +483,27 @@ def output_type(self):
         return rlz.shape_like(tuple(self.flat_args()), dt.string)
 
 
-class RealUnaryOp(UnaryOp):
+class MathUnaryOp(UnaryOp):
     arg = Arg(rlz.numeric)
-    output_type = rlz.shape_like('arg', dt.double)
+
+    def output_type(self):
+        arg = self.arg
+        if isinstance(self.arg, ir.DecimalValue):
+            dtype = arg.type()
+        else:
+            dtype = dt.double
+        return rlz.shape_like(arg, dtype)
```

Comment:
Semantically this is equivalent with 

```py
dtype = dt.highest_precedence([self.arg.type(), dt.double])
```

probably faster though

Reply:
Which reminds me, We might add a `functools.lru_cache` for `dt.highest_precedence`

Reply:
https://github.com/ibis-project/ibis/issues/1545